KAFKA-16013: Add metric for expiration rate of delayed remote fetch

nikramakrishnan opened the pull request:
Add metric for the number of expired remote fetches per second, and corresponding unit test to verify that the metric is marked on expiration.

```
kafka.server:type=DelayedRemoteFetchMetrics,name=ExpiresPerSec
```

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

nikramakrishnan committed:
Add metric for expiration rate of delayed remote fetch

showuon commented:
@nikramakrishnan , great! I created an integration test in replicaManager, we can merge your PR first, and then I added the integration test later.
This is my drafted PR: https://github.com/apache/kafka/compare/trunk...showuon:KAFKA-16013?expand=1
Basically, we're doing the same implementation. Nice! But please check the JIRA assigner next time before starting doing it. :)

showuon commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -80,6 +81,8 @@ class DelayedRemoteFetch(remoteFetchTask: Future[Void],
     val cancelled = remoteFetchTask.cancel(true)
     if (!cancelled) debug(s"Remote fetch task for for RemoteStorageFetchInfo: $remoteFetchInfo could not be cancelled and its isDone value is ${remoteFetchTask.isDone}")
+
+    DelayedRemoteFetchMetrics.expiredRequestMeter.mark()
```
It would be good if we could log warning when following trying to do remote fetch like I did: https://github.com/apache/kafka/compare/trunk...showuon:KAFKA-16013?expand=1#diff-13a87a3a84b41ec66eb8ce8bc46dc166158f268d214821a3e9eaf28029eceef4R82

nikramakrishnan replied:
This makes sense, thanks. I have added the log.

showuon reviewed (commented):
@nikramakrishnan , thanks for the PR. Left a minor comment. Otherwise, LGTM.

nikramakrishnan committed:
Add warning when follower tries to do a remote fetch

nikramakrishnan commented:
Thanks for the quick review, @showuon, and apologies for creating the PR when the JIRA was assigned to you (I already had written a test for it some time back, just did not find the time to push it and create this).

I have addressed your comment, and this is ready for review.

The integration test is a nice addition, thanks!

showuon reviewed (approved):
LGTM! Thanks for the PR!

showuon commented:
@kamalcph @clolov , do you want to have another review?

clolov commented on core/src/test/scala/integration/kafka/server/DelayedRemoteFetchTest.scala:
```diff
@@ -143,6 +144,52 @@ class DelayedRemoteFetchTest {
+    assertTrue(delayedRemoteFetch.isCompleted)
+
+    // Check that the ExpiresPerSec metric was incremented.
+    val metrics = KafkaYammerMetrics.defaultRegistry.allMetrics
+    assertEquals(1, metrics.keySet.asScala.count(_.getMBeanName == "kafka.server:type=DelayedRemoteFetchMetrics,name=ExpiresPerSec"))
```
For the sake of completeness could you assert that the same metric was 0 or not present prior to this?

nikramakrishnan replied:
Good shout. Also discovered that the replicaId was set to that of a follower (>0). Fixed both and updated. Thanks!

clolov reviewed (approved):
Overall looks good to me - I have just a single request for completeness, but I do not mind it being addressed in a subsequent pull request either by Luke or Nikhil

nikramakrishnan committed:
Assert metric is 0 initially and clean up

kamalcph commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -80,6 +81,11 @@ class DelayedRemoteFetch(remoteFetchTask: Future[Void],
     // cancel the remote storage read task, if it has not been executed yet
     val cancelled = remoteFetchTask.cancel(true)
     if (!cancelled) debug(s"Remote fetch task for for RemoteStorageFetchInfo: $remoteFetchInfo could not be cancelled and its isDone value is ${remoteFetchTask.isDone}")
```
Can we update log statement? There is two `for`

nikramakrishnan replied:
Done, thanks!

kamalcph reviewed (approved):
LGTM!

nikramakrishnan committed:
Minor: Fix log line

satishd commented on core/src/main/scala/kafka/server/DelayedRemoteFetch.scala:
```diff
@@ -79,7 +80,12 @@ class DelayedRemoteFetch(remoteFetchTask: Future[Void],
-    if (!cancelled) debug(s"Remote fetch task for for RemoteStorageFetchInfo: $remoteFetchInfo could not be cancelled and its isDone value is ${remoteFetchTask.isDone}")
+    if (!cancelled) debug(s"Remote fetch task for RemoteStorageFetchInfo: $remoteFetchInfo could not be cancelled and its isDone value is ${remoteFetchTask.isDone}")
+
+    if (fetchParams.isFromFollower)
```
I understand it is a defensive check and this instance is never created if it is a follower fetch. If we want a defensive check, it should be created when `DelayedRemoteFetch` is created instead of creating it in this callback.

showuon replied:
Agree! Let me fix that in my follow-up PR.

satishd reviewed (approved):
Thanks @nikramakrishnan for the PR. OverallLGTM, left a minor comment.

showuon commented:
Re-running the CI since it was broken yesterday: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15014/

showuon commented:
Failed tests are unrelated.
